conn.go

```diff
@@ -39,36 +45,237 @@ type GrpcConnection struct {
 	id            ConnID
 	ip            Address
 	streamWrapper StreamWrapper
+	stopFlag      int32
+	handler       Handler
 	outChan       chan *innerMessage
 	readChan      chan *pb.Message
 	stopChan      chan struct{}
 	sync.RWMutex
 }
 
-func NewConnection(streamWrapper StreamWrapper) (Connection, error) {
-	panic("implement me w/ test case :-)")
+func NewConnection(ip Address, id ConnID, streamWrapper StreamWrapper) (Connection, error) {
+
+	if streamWrapper == nil {
+		return nil, errors.New("fail to create connection ! : streamWrapper is nil")
+	}
+
+	return &GrpcConnection{
+		id:            id,
+		ip:            ip,
+		streamWrapper: streamWrapper,
+		outChan:       make(chan *innerMessage, 200),
+		readChan:      make(chan *pb.Message, 200),
+		stopChan:      make(chan struct{}, 1),
+	}, nil
 }
 
 func (conn *GrpcConnection) Send(data []byte, typ string, successCallBack func(interface{}), errCallBack func(error)) {
-	panic("implement me w/ test case :-)")
+
+	conn.Lock()
+	defer conn.Unlock()
+
+	message, err := conn.build(typ, data)
+	if err != nil {
+		go errCallBack(errors.New(fmt.Sprintf("fail to build message [%s]", err.Error())))
+		return
+	}
+
+	m := &innerMessage{
+		Message:   &message,
+		OnErr:     errCallBack,
+		OnSuccess: successCallBack,
+	}
+
+	conn.outChan <- m
 }
 
-func (conn *GrpcConnection) Ip() string {
-	panic("implement me w/ test case :-)")
+func (conn *GrpcConnection) Ip() Address {
+	return conn.ip
 }
 
-func (conn *GrpcConnection) Id() string {
-	panic("implement me w/ test case :-)")
+func (conn *GrpcConnection) Id() ConnID {
+	return conn.id
 }
 
 func (conn *GrpcConnection) Close() {
-	panic("implement me w/ test case :-)")
+	if conn.isDie() {
+		return
+	}
+
+	isFisrt := atomic.CompareAndSwapInt32(&conn.stopFlag, int32(0), int32(1))
+
+	if !isFisrt {
+		return
+	}
+
+	conn.stopChan <- struct{}{}
+	conn.Lock()
+
+	conn.streamWrapper.Close()
+
+	conn.Unlock()
 }
 
 func (conn *GrpcConnection) Start() error {
-	panic("implement me w/ test case :-)")
+	errChan := make(chan error, 1)
+
+	go conn.readStream(errChan)
+	go conn.writeStream()
+
+	for !conn.isDie() {
+		select {
+		case stop := <-conn.stopChan:
+			conn.stopChan <- stop
+			return nil
+		case err := <-errChan:
+			return err
+		case message := <-conn.readChan:
+			if conn.verify(message) {
+				if conn.handler != nil {
+					m := Message{Message: message, Conn: conn}
+					conn.handler.ServeRequest(m)
+				}
+			} else {
```

Comment:
plz remove else statement for now

Reply:
Is it necessary to comment about empty `else` ?

Reply:
I don't think so. I think someone who implements `verify` should consider `else` statements